File: src/lib/wow/archive/casc/casc-key.ts
/**
 * Compact in-memory representation of CASC content/encoding keys (16-byte MD5).
 *
 * Stored as a latin1 string (one byte per character) instead of a 32-char hex
 * string, halving map key/value string heap for encoding tables (~2.8M entries).
 */
export type CascKey = string;

/** Convert a CDN/config hex key to compact storage form. */
export function cascKeyFromHex(hex: string): CascKey {
  return Buffer.from(hex, 'hex').toString('latin1');
}

/** Convert a compact key back to lowercase hex for paths, caches, and BLTE hashes. */
export function cascKeyToHex(key: CascKey): string {
  return Buffer.from(key, 'latin1').toString('hex');
}

/** Accept hex (from config) or an already-compact key. */
export function asCascKey(key: string): CascKey {
  if (key.length === 32 && /^[0-9a-f]+$/.test(key)) return cascKeyFromHex(key);
  return key;
}

File: golang/internal/wow/archive/casc/casc_key.go
package casc

import (
	"encoding/hex"
	"regexp"
	"strings"
)

// CascKey is a compact in-memory representation of CASC content/encoding keys.
type CascKey string

var hexKeyPattern = regexp.MustCompile(`^[0-9a-f]+$`)

// CascKeyFromHex converts a CDN/config hex key to compact storage form.
func CascKeyFromHex(hexKey string) CascKey {
	data, _ := hex.DecodeString(hexKey)
	return CascKey(string(data))
}

// CascKeyToHex converts a compact key back to lowercase hex.
func CascKeyToHex(key CascKey) string {
	return hex.EncodeToString([]byte(key))
}

// AsCascKey accepts hex (from config) or an already-compact key.
func AsCascKey(key string) CascKey {
	if len(key) == 32 && hexKeyPattern.MatchString(key) {
		return CascKeyFromHex(key)
	}
	return CascKey(key)
}

// NormalizeHexKey lowercases a hex key for comparisons.
func NormalizeHexKey(key string) string {
	return strings.ToLower(key)
}
